### PR TITLE
Add bagua installation in dockerfile

### LINKED ISSUE #10455: Support Bagua Training Plugin
## 🚀 Feature

Bagua is a deep learning training acceleration framework for PyTorch.
You could read more the supported features and benchmarks [here](https://tutorials.baguasys.com/).
Mnist [Example](https://github.com/BaguaSys/bagua/blob/master/examples/mnist/main.py).

We could support it by introducing a new `BaguaPlugin` for training by passing it to the `strategy` flag

```python
trainer = Trainer(accelerator="gpu", devices=4, strategy="bagua")
```

cc @borda @akihironitta

### PR DESCRIPTION
**Pull request by wangraying**
## What does this PR do?
Fixes #10455

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [n/a] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kaushikb11** [comment]
@wangraying The Script fails on MacOS

**Comment by wangraying** [comment]
> @wangraying The Script fails on MacOS

It seems the failure on MacOS is caused by exceeding max execution time. This script will not affect the workflow of `ci_test-full.yml`.  Would you mind re-running the job?

**Comment by akihironitta** [comment]
Hi @wangraying

> This script will not affect the workflow of `ci_test-full.yml`. Would you mind re-running the job?

You're right. The failing one on macOS is a known issue tracked in #11216. As we can see from the log, the job exceeded the max time, but the actual tests passed. https://github.com/PyTorchLightning/pytorch-lightning/runs/4690064958?check_suite_focus=true#step:16:2867

```
Mon, 03 Jan 2022 10:54:26 GMT === 2245 passed, 509 skipped, 11 xfailed, 3598 warnings in 630.02s (0:10:30) ===
Mon, 03 Jan 2022 11:15:08 GMT Error: The operation was canceled.
```

File: dockers/base-cuda/Dockerfile [hunk 1]
```diff
@@ -114,6 +114,10 @@ RUN \
     # install DeepSpeed
     pip install deepspeed==0.5.4
 
+RUN \
+    # install Bagua
+    pip install bagua-cuda102
```
Comment (justusschock) [inline dockers/base-cuda/Dockerfile:119]: Do you have precompiled versions for different cuda-versions? We need to match this with the system cuda (defined in $CUDA_VERSION)

Comment (wangraying) [reply dockers/base-cuda/Dockerfile:119]: Currently only three cuda-versions (10.2, 11.1 and 11.3) have pre-compiled packages. Otherwise, it will need to be installed from source code. We are discussing about adding for more cuda-versions. Will comment here if I have update. More info is here: https://tutorials.baguasys.com/installation#installation

Comment (wangraying) [reply dockers/base-cuda/Dockerfile]: Late for this update. We can provide pre-compiled versions for CUDA 10.2, 11.1, 11.3 and 11.5(we already did so). They are built upon the [`pytorch/manylinux**`](https://hub.docker.com/search?q=pytorch%2Fmanylinux&type=image) images, it seems images for other cuda-versions are not actively maintained now. Would it be all right? @justusschock

File: dockers/base-cuda/Dockerfile [hunk 2]
```diff
@@ -114,6 +115,10 @@ RUN \
     # install DeepSpeed
     pip install deepspeed==0.5.4
 
+RUN \
+    # install Bagua
+    pip install bagua-cuda${BAGUA_CUDA_VERSION}==0.9.0
```
Comment (akihironitta) [inline dockers/base-cuda/Dockerfile:120]: I think we can get rid of `BAGUA_CUDA_VERSION` above, and do the following:
```suggestion
    pip install bagua-cuda${CUDA_VERSION//.}==0.9.0
```

---

It just removes `.` in the variable:
```console
$ export SOME_VAR="10.2"
$ echo ${SOME_VAR//.}
102
```

Comment (wangraying) [reply dockers/base-cuda/Dockerfile]: Thanks  @akihironitta

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: dockers/base-cuda/Dockerfile [hunk 3]
```diff
@@ -114,6 +114,10 @@ RUN \
     # install DeepSpeed
     pip install deepspeed==0.5.7
 
+RUN \
+    # install Bagua
+    pip install bagua-cuda${CUDA_VERSION//.}==0.9.0
```
Comment (carmocca) [inline dockers/base-cuda/Dockerfile:119]: This did not work:

```python
 > [10/11] RUN     pip install bagua-cuda${CUDA_VERSION//.}==0.9.0:
#13 1.258 ERROR: Could not find a version that satisfies the requirement bagua-cuda10289==0.9.0 (from versions: none)
#13 1.258 ERROR: No matching distribution found for bagua-cuda10289==0.9.0
```

File: dockers/base-cuda/Dockerfile (hunk 1 above)
Comment (tchaton) [reply dockers/base-cuda/Dockerfile]: @justusschock
